### DISCUSSION TITLE
matchFulfilled is not fired for cached results

### DISCUSSION THREAD

**Discussion by tkharuk** [reduxjs/redux-toolkit#1717] [Q&A]
Hi! struggling with a following scenario a bit

- user clicks link **1** from the sidebar
- URL changes, RTK loads config data via `getConfig`
- `extraReducers` listens for `getConfig.matchFulfilled` and saves payload
  - also clears some non-BE but related to config state
- user clicks on  link **2** from sidebar, everything works the same
- user pokes around, non-BE related shared state changes and is written to `slice`
- user clicks on link **1** again
- nothing happens because callback for `getConfig.matchFulfilled` in `extraReducers` is never fired

I would expect `matchFulfilled` to be triggered disregarding the source (either api or cache), or is this expected?
I could invalidate cache at some point, easiest done when user initiates a click on sidebar link I suppose.

also having a hard time finding relevant information 😿 
<img width="579" alt="Screenshot 2021-11-10 at 6 05 25 PM" src="https://user-images.githubusercontent.com/1205052/141148892-a122d171-87a3-4bdf-94b9-2d9a56e57354.png">

**Comment by phryneas** [comment]
`matchFulfilled` is only fired when also a request has been made.
There are three ways a request can take:

* pending -> fulfilled
* pending -> rejected
* pending -> rejected because of a condition (`action.meta.condition` will be `true` - this is a cache hit)

**Reply by tkharuk** [reply]
@phryneas so it is rejected and there is no payload from cache?
So, how to keep in sync state in Query with some related slice in redux?

**Reply by phryneas** [reply]
There is no new data - there is nothing to sync?

**Reply by tkharuk** [reply]
In a scenario described above, I would have a stale state in slice after clicking on "link 1" for the second time

**Reply by phryneas** [reply]
If you keep completely unrelated data in a slice, you will have to somehow normalize that so it does not overwrite each other. By copying data out of RTK-Query you agree that you need to manage that data for yourself - there is not much RTK-Q can help you with that.

**Reply by tkharuk** [reply]
yeah, exactly, I do copy data from RTKQ into slice.
I could sync it via `useEffect`, but than, this is the same I would do with React-Query or any other non-redux library.

Maybe I'm just trying to use the capabilities wrongly.


How would you recommend to do the following?
- user loads url with some `ID`
- fetch some `config(ID)` from BE
- on UI, there are multiple components that read from that `config(ID)` and can change it somehow
  - also changing `config(ID)` marks it as `dirty` - a property that is not part of the BE state
- by clicking on some link url changes to new `ID`
- all previously changed things in `config` should be reseted

**Reply by phryneas** [reply]
I'd track the **diff** in a slice - normalized by entity id. Merge them in a selector.

**Reply by tkharuk** [reply]
once the user click on new `ID`, old one's state data is not relevant anymore, and should be purged anyway.
So, we'd end up with normalized data where only 1 entity present.
We really care only about "current" one

**Reply by tkharuk** [reply]
@phryneas I think I do understand the idea of your approach.
It ensures stability of data.
But it slightly makes app code more complicated for the given UI.

I think I would rather not cache this particular endpoint at all.
What are the simplest ways to do it?

I suppose, this should work, looks a bit weird, not sure it's optimal
```js
 getConfig: query({
      query: (configId) => `config/${configId}`,
      providesTags: ['Config'],
      async onQueryStarted(_id, { dispatch }) {
        dispatch(api.util.invalidateTags(['Config']));
      },
    }),
```

**Reply by phryneas** [reply]
You could also just use `refetchOnMountOrArgChange`.

**Reply by tkharuk** [reply]
`refetchOnMountOrArgChange`can be set globally or on the hook level.
But I have 5 hooks across the app. There is a clear choice where to put it, but it is not exactly "source of control" I'd like to have 

```js
getView: query({
  query: (viewId) => `accounting-v2-reports/views/${viewId}`,
  refetchOnMountOrArgChange: true, // would make more sense to me :(
}),
```

**Reply by georgeiliadis91** [reply]
Honestly, I would expect the cache hit to trigger a matchFullfilled. And have the data included in the payload. 

Is there any change this behavior changes in the future?

**Reply by markerikson** [reply]
No, because there is no behavior to change.

A "cache hit" means that `useSelector(state => state.api.queries[cacheKey])` returned an actual value. There is no action being dispatched when you _read_ from state, therefore there's nothing for `matchFulfilled` to match against.

What you're asking for would involve A) dispatching an action _while a component is rendering_ (against the rules of React), and B) doing so _every time a `useSelector` reads that cache entry_ (hundreds or thousands of times).

**Reply by georgeiliadis91** [reply]
Fair.. but i have to say it is a bit confusing. Since the match rejected is the one that is triggered..

**Reply by markerikson** [reply]
@georgeiliadis91 I don't understand that statement. Can you clarify that? What are you trying to accomplish overall?

**Reply by georgeiliadis91** [reply]
> `matchFulfilled` is only fired when also a request has been made. There are three ways a request can take:
> 
> * pending -> fulfilled
> * pending -> rejected
> * pending -> rejected because of a condition (`action.meta.condition` will be `true` - this is a cache hit)

I am referring to the above, i assume the last scenario means that there is a cache hit. Unless I am miss understand it.

Basically what I have, is a bunch of logic in the middleware.  I want to detect my cache hits inside the middleware, and given the payload do some action.  

I think I can get the cachekey from the matchRejected and thus be able to extract the data from the store. 
But i would expect the cache hit behaviour to be similar to request being successfull, inside the middleware.

**Reply by markerikson** [reply]
@georgeiliadis91 I checked our internal logic around that statement.

The internal query thunks _will_ bail out with that `action.meta.condition: true`  value in a couple cases:

- the thunk was dispatched, but the cache entry already has a status of `'pending'`
- the thunk was dispatched, it's not a refetch case, and we already have an entry with a status of `'fulfilled'`

However, neither of those happens in typical usage.

Can you give an example of where and how you are expecting there to be a cache hit in your application? I get that you're trying to do some work in a middleware, but what are you expecting to _trigger_ a "cache hit" elsewhere in your codebase?

**Reply by georgeiliadis91** [reply]
On a different computer right now, But here is a pseudo snippet, 


```js
switch(true):
 case setFilterA.match(action):
 case setFilterB.match(action):
 case setFilterC.match(action):
 case setFilterD.match(action):
    myapi.endpoints.myendpoint.initiate(valA,valB,valC,valD) 
    return next(action)
 case myapi.endpoints.myendpoint.matchFullfiled(action): 
   dispatchEvent('update array data', action.payload.data)
   return next(action)
 case myapi.endpoints.myendpoint.matchRejected(action): 
   // I need to check here if this was rejected because the request failed, 
   // or because there was a cache hit. I am basically duplicating my logic 
   // for a successfull response received
   
   if(there is a cache hit){
      // Need to get the cache key
      // Find the data and then proceed
    dispatchEvent('update array data', action.payload.data)
   }else{ 
     do error handling here
   }
     
   return next(action) 
   
```

I have a bunch of actions that both initiate a request inside the middleware, that request is expensive.
If I have a cached value of  a given combination of said filters I would like to use the cached value, since it rarely changes. 

But as is the first time, I have to get the fresh request on the matchFulfilled and the ones after on the rejected case. 
Hope this is clear of what I am trying to do. If there is a better approach please let me know.

**Reply by markerikson** [reply]
Off the top of my head I'm not sure you're going to be able to do what you're looking for just based on the thunks.

`createAsyncThunk` is designed such that if the `condition` option callback returns false, it will dispatch the `'rejected'` action.  The action will indicate that it bailed out, but there's no reason given _why_ it chose to do so - all it knows is that `condition` said "don't run the request".

For RTKQ's internal usage of `createAsyncThunk`, the logic looks like this:

```ts
  const queryThunk = createAsyncThunk<
    ThunkResult,
    QueryThunkArg,
    ThunkApiMetaConfig & { state: RootState<any, string, ReducerPath> }
  >(`${reducerPath}/executeQuery`, executeEndpoint, {
    getPendingMeta() {
      return { startedTimeStamp: Date.now(), [SHOULD_AUTOBATCH]: true }
    },
    condition(queryThunkArgs, { getState }) {
      const state = getState()

      const requestState =
        state[reducerPath]?.queries?.[queryThunkArgs.queryCacheKey]
      const fulfilledVal = requestState?.fulfilledTimeStamp
      const currentArg = queryThunkArgs.originalArgs
      const previousArg = requestState?.originalArgs
      const endpointDefinition =
        endpointDefinitions[queryThunkArgs.endpointName]

      // Order of these checks matters.
      // In order for `upsertQueryData` to successfully run while an existing request is in flight,
      /// we have to check for that first, otherwise `queryThunk` will bail out and not run at all.
      if (isUpsertQuery(queryThunkArgs)) {
        return true
      }

      // Don't retry a request that's currently in-flight
      if (requestState?.status === 'pending') {
        return false
      }

      // if this is forced, continue
      if (isForcedQuery(queryThunkArgs, state)) {
        return true
      }

      if (
        isQueryDefinition(endpointDefinition) &&
        endpointDefinition?.forceRefetch?.({
          currentArg,
          previousArg,
          endpointState: requestState,
          state,
        })
      ) {
        return true
      }

      // Pull from the cache unless we explicitly force refetch or qualify based on time
      if (fulfilledVal) {
        // Value is cached and we didn't specify to refresh, skip it.
        return false
      }

      return true
    },
    dispatchConditionRejection: true,
  })
```

We only hit that last case if you explicitly ended up dispatching the same thunk with the same args while there was an existing cache entry already fulfilled, and if that happens all the action shows is `action.meta.condition: true`. There's no reason shown.

That may be enough for your logic to decide "well, I guess that _must_ have been a cache hit". Or it may not.  You might have to also actually run the selector for that endpoint with the same args, find the cache entry, and check its `fulfilled` status.

**Reply by georgeiliadis91** [reply]
> We only hit that last case if you explicitly ended up dispatching the same thunk with the same args while there was an existing cache entry already fulfilled, and if that happens all the action shows is `action.meta.condition: true`. There's no reason shown.
> 
> That may be enough for your logic to decide "well, I guess that _must_ have been a cache hit". Or it may not. You might have to also actually run the selector for that endpoint with the same args, find the cache entry, and check its `fulfilled` status.

Yep, that was what I was planning to do, check the selector if there is a cache key present. It just feels bit funny that the cache hit is matched at the reject and not the fulfilled :/

Are there any plans to implement something like matchCachehit down the roadmap? Or even if it doesn't exist, do you believe it is a good idea? 

Btw @markerikson, thank you for the replies so far! :)

**Reply by markerikson** [reply]
No plans, and no, I don't think it's a good idea.

**Reply by georgeiliadis91** [reply]
Thanks for the reply!
